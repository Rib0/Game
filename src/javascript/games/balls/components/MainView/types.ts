import { CHANGE_FLASKS, CHANGE_ACTIVE_FLASK_ID, CHANGE_TARGET_COORDS, SET_WIN } from './reducer';
import { ballsColors } from './index';

export type activeFlaskIdType = string | null;
export type coordsType = {
    bottom: number;
    left: number;
};

export interface IBall {
    color: ballsColors;
    id: string;
}

export interface flaskType {
    id: string;
    balls: IBall[];
}
export type flasksType = flaskType[];
export interface IinitialState {
    flasks: flasksType;
    activeFlaskId: activeFlaskIdType;
    targetCoords: coordsType | null;
    isWin: boolean;
}

export interface IChangeFlasksAction {
    type: typeof CHANGE_FLASKS;
    payload: flasksType;
}

export interface IChangeActiveFlaskIdAction {
    type: typeof CHANGE_ACTIVE_FLASK_ID;
    payload: activeFlaskIdType;
}

export interface IChangeTargetCoords {
    type: typeof CHANGE_TARGET_COORDS;
    payload: coordsType;
}

export interface ISetWin {
    type: typeof SET_WIN;
    payload: boolean;
}

export type actionTypes =
    | IChangeFlasksAction
    | IChangeActiveFlaskIdAction
    | IChangeTargetCoords
    | ISetWin;
