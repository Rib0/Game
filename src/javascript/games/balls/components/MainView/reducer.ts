import { IinitialState, activeFlaskType, coordsType, flasksType, actionTypes } from './types';

export const CHANGE_FLASKS = 'CHANGE_FLASKS';
export const CHANGE_ACTIVE_FLASK = 'CHANGE_ACTIVE_FLASK';
export const CHANGE_ACTIVE_COORDS = 'CHANGE_ACTIVE_COORDS';
export const CHANGE_TARGET_COORDS = 'CHANGE_TARGET_COORDS';
export const SET_WIN = 'SET_WIN';

export const changeFlasks = (payload: flasksType): actionTypes => ({
    type: CHANGE_FLASKS,
    payload,
});

export const changeActiveFlask = (payload: activeFlaskType): actionTypes => ({
    type: CHANGE_ACTIVE_FLASK,
    payload,
});

export const changeTargetCoords = (payload: coordsType): actionTypes => ({
    type: CHANGE_TARGET_COORDS,
    payload,
});

export const setWin = (payload: boolean): actionTypes => ({
    type: SET_WIN,
    payload,
});

export default (state: IinitialState, action: actionTypes): IinitialState => {
    switch (action.type) {
        case CHANGE_FLASKS:
            return {
                ...state,
                flasks: action.payload,
            };
        case CHANGE_ACTIVE_FLASK:
            return {
                ...state,
                activeFlask: action.payload === state.activeFlask ? null : action.payload,
            };
        case CHANGE_TARGET_COORDS:
            return {
                ...state,
                targetCoords: action.payload && {
                    ...state,
                    ...action.payload,
                },
            };
        case SET_WIN:
            return {
                ...state,
                isWin: action.payload,
            };
        default:
            return state;
    }
};
