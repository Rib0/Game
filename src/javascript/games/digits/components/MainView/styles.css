.value {
    height: 100%;
    padding: 5px;
    font-size: calc(var(--defaultFontSize) * 3);
    display: flex;
    justify-content: center;
    align-items: center;
    font-weight: bold;
    user-select: none;
    cursor: pointer;
}

.button {
    color: red;
    border-color: red;
    margin-left: 50%;
    margin-top: 20;
    transform: translateX(-50%);
}
