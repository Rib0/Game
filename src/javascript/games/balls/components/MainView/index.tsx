import React, { useState, useEffect, useReducer } from 'react';
import styled from 'styled-components';
import { nanoid } from 'nanoid';

import Header from '../Header';
import Flask from '../Flask';
import Button from '../Button';
import Modal from '../Modal';

import reducer, { changeFlasks, changeActiveFlaskId, changeTargetCoords, setWin } from './reducer';
import { IinitialState, flaskType, flasksType } from './types';
import { getRandom, splitArray, isFilledFlask } from '../../utils';

/*
    возможность менять темы
    возможность выбрать уровень сложности
*/

const Container = styled.div`
    padding: 20px;
`;

const GameField = styled.div`
    padding-top: 70px;
    display: flex;
    flex-wrap: wrap;
`;

const Row = styled.div`
    display: flex;
    justify-content: center;
    flex: 0 0 100%;
    margin-bottom: 60px;
`;

export enum ballsColors {
    darkGreen = '#78970e',
    veryDarkGreen = '#116533',
    lightGreen = '#61d67d',
    purple = '#722b93',
    brown = '#7d4a08',
    darkBlue = '#3b2fc3',
    blue = '#55a3e5',
    pink = '#ea5e7b',
    yellow = '#f1da58',
    gray = '#636466',
    red = '#c52b23',
    orange = '#e88c41',
}

export const initialState: IinitialState = {
    // theme: default,
    flasks: [],
    activeFlaskId: null,
    targetCoords: null,
    isWin: false,
};

const FLASKS_AMOUNT = 14;
const BALLS_PER_FLASK = 4;

const MainView = () => {
    const [{ flasks, activeFlaskId, targetCoords, isWin }, dispatch] = useReducer(
        reducer,
        initialState
    );
    const [targetFlaskId, setTargetFlaskId] = useState('');
    const [prevMoveSnapshot, setPrevMoveSnapshot] = useState<flasksType[]>([]);

    const fillFlasks = () => {
        const colorsKeys: ballsColors[] = Object.values(ballsColors);

        const colorsCounter = colorsKeys.reduce((acc, cur) => {
            return {
                ...acc,
                [cur]: 0,
            };
        }, {});

        let resultFlasks: flasksType = new Array(FLASKS_AMOUNT).fill({});

        resultFlasks = resultFlasks.map((f, i) => {
            const flask: flaskType = { id: nanoid(5), balls: [] };

            if (i > FLASKS_AMOUNT - 3) return flask;

            while (flask.balls.length < BALLS_PER_FLASK) {
                const randomColorIndex = getRandom(0, colorsKeys.length - 1);
                const randomColor = colorsKeys[randomColorIndex];
                const colorCount = colorsCounter[randomColor];

                if (colorCount !== BALLS_PER_FLASK) {
                    flask.balls.push({ color: randomColor, id: nanoid(5) });
                    colorsCounter[randomColor] = colorCount + 1;
                }
            }

            return flask;
        });

        dispatch(changeFlasks(resultFlasks));
    };

    const findFlask = (id: string): Partial<flaskType> =>
        flasks.find(flask => flask.id === id) || {};

    const onClick = (e: React.MouseEvent<HTMLDivElement>) => {
        if (!(e.currentTarget instanceof HTMLDivElement)) {
            return;
        }

        const {
            dataset: { id },
        } = e.currentTarget;
        const targetFlask = findFlask(id);
        const isFilled = isFilledFlask(targetFlask.balls);

        if (isFilled || isWin) {
            dispatch(changeActiveFlaskId(null));
            return;
        }

        if (activeFlaskId !== id && activeFlaskId !== null) {
            const [changedBall] = findFlask(activeFlaskId).balls.slice(-1);
            const [topTargetBall] = targetFlask.balls.slice(-1);

            if (topTargetBall && topTargetBall.color !== changedBall.color) {
                dispatch(changeActiveFlaskId(null));
                return;
            }

            const { bottom, left } = e.currentTarget.getBoundingClientRect();
            const restBalls = findFlask(activeFlaskId).balls.slice(0, -1);
            const ballsWithChanged = targetFlask.balls.concat(changedBall);
            const updatedFlasks = flasks.map((flask, i) =>
                flask.id === activeFlaskId
                    ? { ...flask, balls: restBalls }
                    : flask.id === id
                        ? { ...flask, balls: ballsWithChanged }
                        : flask
            );

            dispatch(changeTargetCoords({ bottom, left }));
            setTargetFlaskId(id);
            setPrevMoveSnapshot(spanshots => [flasks, ...spanshots]);

            setTimeout(() => {
                dispatch(changeFlasks(updatedFlasks));
                dispatch(changeActiveFlaskId(null));
                dispatch(changeTargetCoords(null));
            }, 300);
        } else if (targetFlask.balls.length) {
            dispatch(changeActiveFlaskId(id));
        }
    };

    const checkWin = () => {
        const hasWin = flasks
            .filter(({ balls }) => balls.length)
            .every(({ balls }) => isFilledFlask(balls));

        if (hasWin) {
            setWin(true);
        }
    };

    const handleAddFlask = () => {
        if (flasks.length > FLASKS_AMOUNT) return;

        dispatch(changeFlasks(flasks.concat({ id: nanoid(5), balls: [] })));
    };

    const handleCanselLastMoove = () => {
        if (!prevMoveSnapshot.length) return;

        const [lastSnapshot, ...rest] = prevMoveSnapshot;

        dispatch(changeFlasks(lastSnapshot));
        setPrevMoveSnapshot(rest);
    };

    const handleRestart = () => {
        setWin(false);
        fillFlasks();
    };

    useEffect(fillFlasks, []);

    useEffect(checkWin, [flasks]);

    return (
        <Container>
            <Header>
                <Button onClick={handleRestart} icon="sync-alt" />
                <Button onClick={handleCanselLastMoove} icon="reply" />
                <Button
                    disabled={flasks.length > FLASKS_AMOUNT}
                    onClick={handleAddFlask}
                    icon="plus"
                />
            </Header>
            <GameField>
                {splitArray(flasks, FLASKS_AMOUNT / 2, FLASKS_AMOUNT < flasks.length).map(
                    (row, rowIndex) => (
                        <Row key={rowIndex}>
                            {row.map(flask => {
                                const isActive = flask.id === activeFlaskId;
                                const targetBallsLength = findFlask(targetFlaskId).balls?.length;

                                return (
                                    <Flask
                                        key={flask.id}
                                        onClick={onClick}
                                        balls={flask.balls}
                                        flaskId={flask.id}
                                        isActive={isActive}
                                        targetCoords={targetCoords}
                                        targetBallsLength={targetBallsLength}
                                        flasksLength={flasks.length}
                                    />
                                );
                            })}
                        </Row>
                    )
                )}
            </GameField>
            <Modal
                header="Вы выиграли!"
                content={<>Хотите сыграть еще?</>}
                isActive={isWin}
                acceptButtonText="Начать заного"
                acceptButtonCallback={handleRestart}
                declineButtonText="Вернуться в главное меню"
                declineButtonCallback={(window as any)?.backToMenu}
            />
        </Container>
    );
};

export default MainView;
