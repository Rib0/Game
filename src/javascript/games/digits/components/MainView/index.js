import React, { PureComponent } from 'react';

import Field from '../Field';

import styles from './styles.css';

class MainView extends PureComponent {
    static SideSize = new Array(4);

    static CellSize = 100;

    state = {
        field: [],
        loading: true,
    };

    getRandom(min, max) {
        min = Math.ceil(min);
        max = Math.floor(max);
        return Math.floor(Math.random() * (max - min)) + min;
    }

    generateField = () => {
        this.setState({ loading: true });
        setTimeout(() => {
            this.loadCells();
            setTimeout(() => this.setState({ loading: false }), 0);
        }, 1000);
    };

    loadCells() {
        const fieldsCount = Math.pow(MainView.SideSize.length, 2);

        const field = new Array(fieldsCount - 1).fill(1).reduce((acc, v, index) => {
            let randomValue = this.getRandom(1, fieldsCount);
            while (acc.find(({ value }) => value === randomValue)) {
                randomValue = this.getRandom(1, fieldsCount);
            }
            acc.push({
                value: randomValue,
                order: index + 1,
            });

            return acc;
        }, []);

        this.setState({
            field,
        });
    }

    getCellStyles(index) {
        const { loading } = this.state;
        const sideSize = MainView.SideSize.length;
        const leftMulti = index % sideSize;
        const topMulti = parseInt(index / sideSize);

        return {
            top: loading ? 0 : `${topMulti * MainView.CellSize}px`,
            left: loading ? 0 : `${leftMulti * MainView.CellSize}px`,
            opacity: index < Math.pow(sideSize, 2) - 1 && loading ? 0 : 1,
        };
    }

    findEmptyCell(index) {
        const { field } = this.state;
        const fieldsCount = Math.pow(MainView.SideSize.length, 2);

        const emptyCell = Array.from(new Array(fieldsCount).keys()).find(
            value => !field.map(({ order }) => order).includes(value)
        );
        const emptyCellIndex = [index - 1, index + 1, index - 4, index + 4].find(
            index => index === emptyCell
        );

        return emptyCellIndex;
    }

    onClick = e => {
        const {
            dataset: { index },
        } = e.currentTarget;
        const emptyIndex = this.findEmptyCell(Number(index));
        if (emptyIndex === undefined) return;

        const { field } = this.state;
        const newField = field.map(cell =>
            cell.order === Number(index) ? { ...cell, order: emptyIndex } : cell
        );

        this.setState({
            field: newField,
        });
    };

    render() {
        const { field } = this.state;

        return (
            <>
                <button onClick={this.generateField}>Сгенерировать</button>
                <Field>
                    {field.map(({ value, order }) => (
                        <Field.Cell
                            style={this.getCellStyles(order)}
                            onClick={this.onClick}
                            data={{
                                'data-index': order,
                            }}
                            key={value}
                        >
                            <div className={styles.value}>{value}</div>
                        </Field.Cell>
                    ))}
                </Field>
            </>
        );
    }
}

export default MainView;
