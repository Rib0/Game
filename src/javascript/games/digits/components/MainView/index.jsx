import React, { PureComponent } from 'react';

import { Button, Modal } from 'Base';
import Field from '../Field';

import styles from './styles.css';

class MainView extends PureComponent {
    static SideSize = new Array(4);
    static CellSize = 100;
    static defaultLoadTime = 500;

    state = {
        field: [],
        loading: false,
        hasWin: false,
    };

    componentDidUpdate(p, prevState) {
        const { prevField } = prevState;
        const { field, loading } = this.state;

        if (prevField !== field && !loading) {
            this.checkWin();
        }
    }

    getRandom(min, max) {
        min = Math.ceil(min);
        max = Math.floor(max);
        return Math.floor(Math.random() * (max - min)) + min;
    }

    loadCells() {
        const fieldsCount = MainView.SideSize.length ** 2;

        const field = new Array(fieldsCount - 1).fill(1).reduce((acc, v, index) => {
            let randomValue = this.getRandom(1, fieldsCount);
            while (acc.find(({ value }) => value === randomValue)) {
                randomValue = this.getRandom(1, fieldsCount);
            }
            acc.push({
                value: randomValue,
                order: index + 1,
            });

            return acc;
        }, []);

        this.setState({
            field,
        });
    }

    generateField = () => {
        const { loading } = this.state;
        if (loading) return;

        this.setState({ loading: true });
        setTimeout(() => {
            this.loadCells();
            setTimeout(() => this.setState({ loading: false }), 0);
        }, MainView.defaultLoadTime);
    };

    getCellStyles(index) {
        const { loading } = this.state;
        const sideSize = MainView.SideSize.length;
        const leftMulti = index % sideSize;
        const topMulti = parseInt(index / sideSize, 10);

        return {
            top: loading ? 0 : `${topMulti * MainView.CellSize}px`,
            left: loading ? 0 : `${leftMulti * MainView.CellSize}px`,
            opacity: index < sideSize ** 2 - 1 && loading ? 0 : 1,
        };
    }

    findEmptyCell(index) {
        const { field } = this.state;
        const fieldsCount = MainView.SideSize.length ** 2;

        const emptyCell = Array.from(new Array(fieldsCount).keys()).find(
            value => !field.map(({ order }) => order).includes(value)
        );
        const emptyCellIndex = [index - 1, index + 1, index - 4, index + 4].find(
            index => index === emptyCell
        );

        return emptyCellIndex;
    }

    // makeWin = () => { for test only
    //     const { field } = this.state;

    //     this.setState({
    //         loading: true,
    //     });

    //     setTimeout(() => {
    //         const sortedField = field.map((v, index) => ({ value: index + 1, order: index }));
    //         this.setState({ field: sortedField });
    //         setTimeout(() => {
    //             this.setState({ loading: false });
    //         }, 0);
    //     }, MainView.defaultLoadTime);
    // };

    checkWin() {
        const { field } = this.state;
        const hasWin = field.every(({ value, order }, index) => value === order + 1);

        if (hasWin) {
            this.setState({
                hasWin,
            });
        }
    }

    onAcceptWin = () => {
        this.setState({ hasWin: false });

        this.generateField();
    };

    onClick = e => {
        const {
            dataset: { index },
        } = e.currentTarget;
        const emptyIndex = this.findEmptyCell(Number(index));
        if (emptyIndex === undefined) return;

        const { field } = this.state;
        const newField = field.map(cell =>
            cell.order === Number(index) ? { ...cell, order: emptyIndex } : cell
        );

        this.setState({
            field: newField,
        });
    };

    render() {
        const { field, loading, hasWin } = this.state;

        return (
            <div className={styles.container}>
                <Button
                    size={Button.Sizes.large}
                    className={styles.button}
                    onClick={this.generateField}
                    disabled={loading}
                    text="Сгенерировать"
                />
                <Field>
                    {field.map(({ value, order }) => {
                        const style = {
                            ...this.getCellStyles(order),
                            backgroundColor: value - order === 1 && 'lightgreen',
                        };

                        return (
                            <Field.Cell
                                style={style}
                                onClick={this.onClick}
                                data={{
                                    'data-index': order,
                                }}
                                key={value}
                            >
                                <div className={styles.value}>{value}</div>
                            </Field.Cell>
                        );
                    })}
                </Field>
                <Modal
                    isOpen={hasWin}
                    caption="Вы выиграли!"
                    acceptButtonText="Начать заного"
                    onAccept={this.onAcceptWin}
                >
                    <Modal.Content>Игра окончена</Modal.Content>
                </Modal>
            </div>
        );
    }
}

export default MainView;
