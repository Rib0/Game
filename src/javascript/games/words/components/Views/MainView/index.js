import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { VIEWS } from 'games/words/utils/constants';
import styles from './styles.css';

class MainView extends PureComponent {
    render() {
        const { view } = this.props;
        const { component: Component, props } = VIEWS[view];

        return (
            <div className={styles.container}>
                <Component {...props} />
            </div>
        );
    }
}

const mapStateToProps = ({ view }) => ({
    view,
});

MainView.propTypes = {
    view: PropTypes.oneOf(Object.keys(VIEWS)),
};

export default connect(mapStateToProps)(MainView);
