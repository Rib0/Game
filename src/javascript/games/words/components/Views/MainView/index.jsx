import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { VIEWS } from 'WordsGame/utils/constants';
import { dropStore } from 'WordsGame/store/actions';
import styles from './styles.css';

class MainView extends PureComponent {
    componentWillUnmount() {
        const { dropStore } = this.props;
        dropStore();
    }

    render() {
        const { view } = this.props;
        const { component: Component, props } = VIEWS[view];

        return (
            <div className={styles.container}>
                <Component {...props} />
            </div>
        );
    }
}

const mapStateToProps = ({ view }) => ({
    view,
});

const mapDispatchToProps = {
    dropStore,
};

MainView.propTypes = {
    view: PropTypes.oneOf(Object.keys(VIEWS)),
    dropStore: PropTypes.func
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(MainView);
