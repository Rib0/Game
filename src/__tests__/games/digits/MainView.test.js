import React from 'react';
import MainView from 'games/digits/components/MainView';

jest.useFakeTimers();
const setUp = () => shallow(<MainView />);

describe('Digits game - MainView test', () => {
    let Component, instance;

    beforeEach(() => {
        Component = setUp();
        instance = Component.instance();
    });
    afterEach(() => {
        jest.useRealTimers();
    });

    test('Should render', () => {
        expect(Component).toMatchSnapshot();
    });
    test('LoadCells method should generate field array', () => {
        instance.loadCells();
        expect(Component.state().field).toHaveLength(15);
    });
    test('GenerateField method should generate field array', () => {
        jest.useFakeTimers('legacy');
        expect(Component.state().loading).toBeFalsy();
        instance.generateField();
        expect(Component.state().loading).toBeTruthy();
        expect(setTimeout).toHaveBeenCalledTimes(1);
        jest.runAllTimers();
        expect(Component.state().loading).toBeFalsy();
        expect(Component.state().field).toHaveLength(15);
    });
    test('GenerateField method should not generate field array', () => {
        const generateField = jest.spyOn(instance, 'generateField');
        instance.setState({ loading: true });
        instance.generateField();
        expect(generateField).toHaveReturned();
        generateField.mockRestore();
    });
    test('Should find empty cell', () => {
        instance.loadCells();
        const {
            field: [{ order: firstFieldOrder }],
        } = Component.state();
        const emptyCellIndex = instance.findEmptyCell(firstFieldOrder);
        expect(emptyCellIndex).toBe(0);
    });
    test('Should accept win', () => {
        const generateField = jest.spyOn(instance, 'generateField');
        instance.onAcceptWin();
        expect(Component.state().hasWin).toBeFalsy();
        expect(generateField).toHaveBeenCalledTimes(1);
        generateField.mockRestore();
    });
    test('Should change cell order if clicked', () => {
        instance.loadCells();
        instance.onClick({ currentTarget: { dataset: { index: '1' } } });
        const [{ order: firstCellOrder }] = Component.state().field;
        expect(firstCellOrder).toBe(0);
    });
    test('Should not change cell order if clicked', () => {
        instance.loadCells();
        instance.onClick({ currentTarget: { dataset: { index: '2' } } });
        const [{ order: firstCellOrder }] = Component.state().field;
        expect(firstCellOrder).toBe(1);
    });
    test('Should render Modal if player wins', () => {
        instance.setState({ hasWin: true });
        expect(Component.find('Modal')).toHaveLength(1);
    });
});
