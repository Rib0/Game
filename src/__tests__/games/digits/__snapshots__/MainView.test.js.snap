// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Digits game - MainView test Should render 1`] = `
<div
  className="container"
>
  <Button
    className="button"
    disabled={false}
    onClick={[Function]}
    size="large"
    text="Сгенерировать"
    type="styled"
  />
  <Field
    field={Array []}
    loading={false}
    onClick={[Function]}
  />
  <Modal
    acceptButtonText="Начать заного"
    caption="Вы выиграли!"
    isOpen={false}
    onAccept={[Function]}
  >
    <Content>
      Игра окончена
    </Content>
  </Modal>
</div>
`;
